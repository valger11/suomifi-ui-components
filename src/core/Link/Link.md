```js
import { Link, Paragraph } from 'suomifi-ui-components';

<Paragraph>
  <Link
    className="test-classname"
    href="https://www.notvisitedlink.com/"
  >
    Not visited link
  </Link>{' '}
  <Link className="test-classname" href="#">
    Visited link
  </Link>
</Paragraph>;
```

### Skip link

Show link for keyboard navigation.

- Should be the first element that can gain focus on the page with the tab.
- The target container should have `tabIndex={-1}` to handle better the functionality with screenreaders.
- Because of the tabIndex you might want set the style of `outline=none` to the target container.
- To see the component, click below on the dashed area and then press `Tab`.

```js
import { LinkSkip } from 'suomifi-ui-components';
<div>
  <div
    style={{
      height: '80px',
      width: '210px',
      borderStyle: 'dashed',
      borderColor: '#000',
      borderWidth: '1px'
    }}
  >
    <LinkSkip href="#">Skip to main content</LinkSkip>
  </div>
</div>;
```

### External link

```js
import { LinkExternal } from 'suomifi-ui-components';

<LinkExternal
  href="https://www.sweden.se/"
  labelNewWindow="Opens to a new window"
>
  External link
</LinkExternal>;
```

### Change component for the link

```js
import { Link } from 'suomifi-ui-components';

const Component = ({ children, ...passProps }) => (
  <a {...passProps}>foo {children} bar</a>
);

<Link
  className="test-classname"
  href="https://www.com/"
  as={Component}
>
  Testing
</Link>;
```
