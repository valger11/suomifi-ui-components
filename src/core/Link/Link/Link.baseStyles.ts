import { css } from 'styled-components';
import { baseStyles } from '../BaseLink/BaseLink.baseStyles';

export const linkExternalStyles = css`
  ${baseStyles}
`;
