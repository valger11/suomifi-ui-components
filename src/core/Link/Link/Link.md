```js
import { Link, Paragraph } from 'suomifi-ui-components';

<Paragraph>
  <Link
    className="test-classname"
    href="https://www.notvisitedlink.com/"
  >
    Not visited link
  </Link>{' '}
  <Link className="test-classname" href="#">
    Visited link
  </Link>
</Paragraph>;
```

### Skip link

Show link for keyboard navigation.

- Should be the first element that can gain focus on the page with the tab.
- The target container should have `tabIndex={-1}` to handle better the functionality with screenreaders.
- Because of the tabIndex you might want set the style of `outline=none` to the target container.
- To see the component, click below on the dashed area and then press `Tab`.

```js
import { SkipLink } from 'suomifi-ui-components';
<div>
  <div
    style={{
      height: '80px',
      width: '210px',
      borderStyle: 'dashed',
      borderColor: '#000',
      borderWidth: '1px'
    }}
  >
    <SkipLink href="#">Skip to main content</SkipLink>
  </div>
</div>;
```

### External link

```js
import { ExternalLink } from 'suomifi-ui-components';

<ExternalLink
  href="https://www.sweden.se/"
  labelNewWindow="Opens to a new window"
>
  External link
</ExternalLink>;
```

### Change component for the link

```js
import { Link } from 'suomifi-ui-components';

const Component = ({ children, ...passProps }) => (
  <a {...passProps}>foo {children} bar</a>
);

<Link
  className="test-classname"
  href="https://www.com/"
  as={Component}
>
  Testing
</Link>;
```
