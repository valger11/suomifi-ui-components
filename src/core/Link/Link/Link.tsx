import React, { Component } from 'react';
import { default as styled } from 'styled-components';
import classnames from 'classnames';
import { baseStyles } from '../BaseLink/BaseLink.baseStyles';
import { HtmlA } from '../../../reset';
import { BaseLinkProps, baseClassName } from '../BaseLink/BaseLink';

export interface LinkProps extends BaseLinkProps {}

const StyledLink = styled(({ asProp, className, ...passProps }: LinkProps) => (
  <HtmlA
    {...passProps}
    className={classnames(baseClassName, className)}
    as={asProp}
  />
))`
  ${baseStyles};
`;

/**
 * <i class="semantics" />
 * Used for adding a link
 */
export class Link extends Component<LinkProps> {
  render() {
    return <StyledLink {...this.props} />;
  }
}
