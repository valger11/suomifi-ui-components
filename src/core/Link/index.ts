export { Link, LinkProps } from './Link/Link';
export { LinkExternal, LinkExternalProps } from './LinkExternal/LinkExternal';
export { LinkSkip, LinkSkipProps } from './LinkSkip/LinkSkip';
