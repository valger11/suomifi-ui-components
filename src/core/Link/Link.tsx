import React, { Component, ReactNode } from 'react';
import { default as styled } from 'styled-components';
import { asPropType } from '../../utils/typescript';
import classnames from 'classnames';
import { baseStyles } from './Link.baseStyles';
import { HtmlA, HtmlAProps } from '../../reset';

export const baseClassName = 'fi-link';

type LinkVariant = 'default' | 'external' | 'skip';
export interface LinkProps extends HtmlAProps {
  /** Link url. Link is not focusable without the href */
  href: string;
  /** Custom classname to extend or customize */
  className?: string;
  /**
   * Link element displayed content
   */
  children: ReactNode;
  variant?: LinkVariant;
  asProp?: asPropType;
}

class BaseLink extends Component<LinkProps> {
  render() {
    const { className, ...passProps } = this.props;
    return (
      <HtmlA
        {...passProps}
        className={classnames(className, {
          [baseClassName]: !className || !className.includes(baseClassName),
        })}
      />
    );
  }
}

const StyledLink = styled(({ asProp, ...passProps }: LinkProps) => (
  <BaseLink {...passProps} as={asProp} />
))`
  ${baseStyles};
`;

/**
 * <i class="semantics" />
 * Used for adding a link
 */
export class Link extends Component<LinkProps> {
  render() {
    return <StyledLink {...this.props} />;
  }
}
