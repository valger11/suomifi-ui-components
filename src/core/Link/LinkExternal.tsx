import React, { Component } from 'react';
import { default as styled } from 'styled-components';
import { logger } from '../../utils/logger';
import classnames from 'classnames';
import { Icon } from '../Icon/Icon';
import { Link, LinkProps } from './Link';
import { VisuallyHidden } from '../VisuallyHidden/VisuallyHidden';
import { externalStyles } from './Link.baseStyles';

const iconClassName = 'fi-link_icon';
const externalClassName = 'fi-link--external';

export interface LinkExternalProps extends LinkProps {
  /** Translated explanation of 'opens to a new window' */
  labelNewWindow: string;
  /** Hide the icon */
  hideIcon?: boolean;
  /** Open to a new window
   * @default true
   */
  toNewWindow?: boolean;
}

class BaseLinkExternal extends Component<LinkExternalProps> {
  render() {
    const { className, toNewWindow = true, ...passProps } = this.props;
    return (
      <Link
        {...passProps}
        className={classnames(className, externalClassName)}
        target={!!toNewWindow ? '_blank' : undefined}
        rel={!!toNewWindow ? 'noopener' : undefined}
      />
    );
  }
}

const StyledLinkExternal = styled(
  (props: Omit<LinkExternalProps, 'labelNewWindow' | 'hideIcon'>) => (
    <Link {...props} asProp={BaseLinkExternal} />
  ),
)`
  ${externalStyles};
`;

/**
 * <i class="semantics" />
 * Used for adding a external site link
 */
export class LinkExternal extends Component<LinkExternalProps> {
  render() {
    const { children, labelNewWindow, hideIcon, ...passProps } = this.props;
    if (!labelNewWindow) {
      logger.warn(
        'External link needs a translated description of link opening to a new window',
      );
    }
    return (
      <StyledLinkExternal {...passProps}>
        {children}
        <VisuallyHidden>{labelNewWindow}</VisuallyHidden>
        {!hideIcon && (
          <Icon
            icon="linkExternal"
            className={iconClassName}
            color="currentColor"
          />
        )}
      </StyledLinkExternal>
    );
  }
}
