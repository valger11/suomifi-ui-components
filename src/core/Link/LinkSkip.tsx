import React, { Component } from 'react';
import classnames from 'classnames';
import { default as styled } from 'styled-components';
import { Link, LinkProps } from './Link';
import { skipLinkStyles } from './Link.baseStyles';

const skipClassName = 'fi-link--skip';

export interface LinkSkipProps extends LinkProps {}

const StyledLinkSkip = styled((props: LinkSkipProps) => <Link {...props} />)`
  ${skipLinkStyles}
`;

/**
 * <i class="semantics" />
 * Used for adding skip link for keyboard and screenreader users
 */
export class LinkSkip extends Component<LinkSkipProps> {
  render() {
    const { className, ...passProps } = this.props;
    return (
      <StyledLinkSkip
        {...passProps}
        className={classnames(className, skipClassName)}
      />
    );
  }
}
