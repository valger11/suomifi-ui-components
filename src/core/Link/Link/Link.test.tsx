import React from 'react';
import { render } from '@testing-library/react';
import { axeTest } from '../../../utils/test/axe';

import { Link } from './Link';

const TestLink = (
  <Link href="/" data-testid="test-link">
    Hey this is test
  </Link>
);

test('calling render with the same component on the same container does not remount', () => {
  const LinkRendered = render(TestLink);
  const { getByTestId, container } = LinkRendered;
  expect(container.firstChild).toMatchSnapshot();
  expect(getByTestId('test-link').textContent).toBe('Hey this is test');
});

test('should not have basic accessibility issues', axeTest(TestLink));
