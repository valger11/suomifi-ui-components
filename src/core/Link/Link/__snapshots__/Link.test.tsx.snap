// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`calling render with the same component on the same container does not remount 1`] = `
.c0 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  background-color: transparent;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  display: inline;
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.c0::before,
.c0::after {
  box-sizing: border-box;
}

.c0:hover,
.c0:active,
.c0:focus,
.c0:focus-within {
  color: inherit;
  -webkit-text-decoration: underline;
  text-decoration: underline;
  cursor: pointer;
}

.c1 {
  color: hsl(0,0%,16%);
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 18px;
  line-height: 1.5;
  font-weight: 400;
  color: hsl(212,63%,45%);
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c1:hover,
.c1:active,
.c1:focus,
.c1:focus-within {
  color: hsl(212,63%,45%);
}

.c1:focus,
.c1:focus-within {
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c1:focus {
  outline: 0;
  border-radius: 2px;
  box-shadow: 0 0 0 2px hsl(196,77%,44%);
}

.c1:hover,
.c1:active {
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.c1:visited {
  color: hsl(284,36%,45%);
}

<a
  class="c0 fi-link c1"
  data-testid="test-link"
  href="/"
>
  Hey this is test
</a>
`;
